### PR TITLE
View Op - new unit tests and add support for tensor memcpy by offset/size

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: Address PR comments

**Motivation and Context**
- Add unit tests for View op
- Support data copy from buffer offset
- allocatormgr.cc code clean up

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/optimizer/common.h [hunk 1]
```diff
@@ -18,5 +18,13 @@ Status CopyIfNotSameBuffer(const Tensor& source_tensor, Tensor& target_tensor) {
   return Status::OK();
 }
 
+template <typename T>
+Status CopyDataFromBufferOffset(const Tensor& source_tensor, Tensor& target_tensor, size_t ptr_offset, size_t count) {
+  const T* source = source_tensor.template Data<T>();
+  T* target = target_tensor.template MutableData<T>();
+  CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target, source + ptr_offset, count, cudaMemcpyDeviceToDevice));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.h:25]: >source + ptr_offset [](start = 47, length = 19)

source can be casted to void* 
and we can directly use byte_offset here

File: orttraining/orttraining/training_ops/cuda/optimizer/common.h [hunk 2]
```diff
@@ -18,5 +18,13 @@ Status CopyIfNotSameBuffer(const Tensor& source_tensor, Tensor& target_tensor) {
   return Status::OK();
 }
 
+template <typename T>
+Status CopyDataFromBufferOffset(const Tensor& source_tensor, Tensor& target_tensor, size_t ptr_offset, size_t count) {
+  const T* source = source_tensor.template Data<T>();
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.h:23]: const void* X_data = X->DataRaw();

File: orttraining/orttraining/training_ops/cuda/optimizer/common.h [hunk 3]
```diff
@@ -18,5 +18,13 @@ Status CopyIfNotSameBuffer(const Tensor& source_tensor, Tensor& target_tensor) {
...
+  T* target = target_tensor.template MutableData<T>();
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.h:24]: >target [](start = 5, length = 6)

can also take raw void* pointer here

File: orttraining/orttraining/training_ops/cuda/tensor/view.cc [hunk 4]
```diff
@@ -68,14 +69,26 @@ Status View::ComputeInternal(OpKernelContext* context) const {
...
+  MLDataType X_type = X->DataType();
   for (int i = 0; i < view_count; ++i) {
     // Outputs are allowed to be unused.
     Tensor* Y = context->Output(i, y_shapes[i]);
     if (Y != nullptr) {
       if (X_data != Y->MutableDataRaw()) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "View output is not sharing the underlaying buffer of input");
+        // View output is not sharing the underlaying buffer of input, copy instead
+        size_t ptr_offset = y_byte_offsets[i] / bytes_per_elem;
+        size_t byte_count = y_shapes[i].Size() * bytes_per_elem;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/tensor/view.cc:80]: >y_shapes[i].Size() * bytes_per_elem; [](start = 28, length = 36)

Y->SizeInBytes()

**Comment by SherlockNoMad** [comment]
Let's more descriptive PR name instead of "PR comments"

File: orttraining/orttraining/training_ops/cuda/tensor/view.cc [hunk 5]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 
 #include "orttraining/training_ops/cuda/tensor/view.h"
+#include "core/providers/cuda/cuda_common.h"
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/tensor/view.cc:5]: not needed anymore?

File: orttraining/orttraining/training_ops/cuda/tensor/view.cc [hunk 6]
```diff
@@ -73,9 +74,14 @@ Status View::ComputeInternal(OpKernelContext* context) const {
     Tensor* Y = context->Output(i, y_shapes[i]);
     if (Y != nullptr) {
       if (X_data != Y->MutableDataRaw()) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "View output is not sharing the underlaying buffer of input");
+        // View output is not sharing the underlaying buffer of input, copy instead
+        const void* source = static_cast<const char*>(X_data) + y_byte_offsets[i];
+        void* target = Y->MutableDataRaw();
+        CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target, source, Y->SizeInBytes(), cudaMemcpyDeviceToDevice));
+        Y->SetByteOffset(0);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/tensor/view.cc:81]: I think this line can be skipped.

Comment (ytaous) [reply orttraining/orttraining/training_ops/cuda/tensor/view.cc:81]: will check

**Pull request merged**